src/client/datascience/notebook/notebookControllerManager.ts

```diff
@@ -0,0 +1,277 @@
+        }
+
+        // Map KernelConnectionMetadata => NotebookController
+        const controllers = kernelConnections.map(value => {
+            return this.createNotebookController(document, value);
+        });
+
+        // Store our NotebookControllers to dispose on doc close
+        this.controllerMapping.set(document, { selected: undefined, controllers: controllers });
+
+        return controllers;
+    }
+
+    private createNotebookController(document: NotebookDocument, kernelConnection: KernelConnectionMetadata): VSCodeNotebookController {
+        // Create notebook selector
+        const controller = new VSCodeNotebookController(document, kernelConnection,
+            this.notebook, this.commandManager, this.kernelProvider,
+            this.preferredRemoteKernelIdProvider, this.context, this.disposables);
+
+        // Hook up to if this NotebookController is selected or de-selected
+        controller.onNotebookControllerSelected(this.handleOnNotebookControllerSelected, this, this.disposables);
+
+        // We are disposing as documents are closed, but do this as well
+        this.disposables.push(controller);
+
+        return controller;
+    }
+
+    // A new NotebookController has been selected, find the associated notebook document and update it
+    private async handleOnNotebookControllerSelected(event: { notebook: NotebookDocument, controller: VSCodeNotebookController }) {
+        if (this.controllerMapping.has(event.notebook)) {
+            const currentMapping = this.controllerMapping.get(event.notebook);
+            // ! Ok here as we have already checked has above
+            this.controllerMapping.set(event.notebook, { controllers: currentMapping?.controllers!, selected: event.controller })
+
+            // Now actually handle the change
+            await this.notebookKernelChanged(event.notebook, event.controller);
+
+            // Now notify out that we have updated a notebooks controller
+            this._onNotebookControllerSelected.fire(event);
+        }
+    }
+
+    // For the given NotebookDocument find all associated KernelConnectionMetadata
+    private async getKernelConnectionMetadata(document: NotebookDocument): Promise<KernelConnectionMetadata[]> {
+        // IANHU: Need a token passed in here?
+        const token = new CancellationTokenSource().token;
+
+        let kernels: KernelConnectionMetadata[] = [];
+        let preferred: KernelConnectionMetadata | undefined;
+
+        // If we already have a kernel selected, set that one as preferred
+        if (this.controllerMapping.has(document)) {
+            preferred = this.controllerMapping.get(document)?.selected?.connection;
+        }
+
+        if (this.isLocalLaunch) {
+            kernels = await this.localKernelFinder.listKernels(document.uri, token);
+            preferred =
+                preferred ??
+                (await this.localKernelFinder.findKernel(document.uri, getNotebookMetadata(document), token));
```

Comment:
I'd change this so that both listKernels and findKernel happens together.
Right now we first list all kernels, wait for that to compelte, then find (if no preferred kernel) the kernel & wait for that to complete

Considering the fact that these are slow & theres a lot of IO (file & process), i think we should parallelize them.

const preferredKernelPromise = preferred ? Promise.resolve(preferred) : this.localKernelFinder.findKernel(document.uri, getNotebookMetadata(document), token);
kernels = await this.localKernelFinder.listKernels(document.uri, token);
preferred = await preferredKernelPromise; // At this point, the promise has already started and waiting on this fater listing kernles doens't matter, as both are running together in parallel